acceptancetests/jujupy/stream_server.py

```diff
@@ -0,0 +1,293 @@
+            '%a, %d %b %Y %H:%M:%S +0000')
+        data = {'updated': updated, 'datatype': 'content-download'}
+        trees = items2content_trees(items, data)
+        return write_juju_streams(self._stream_path, trees, updated)
+
+
+class StreamServer:
+    """Provide service to create stream metadata and to serve it."""
+
+    def __init__(self, base_dir, stream_data_type=_JujuStreamData):
+        self.base_dir = base_dir
+        self.stream_data = stream_data_type(base_dir)
+
+    def add_product(self, content_id, version, arch, series, agent_tgz_path):
+        """Add a new product to generate stream data for.
+
+        :param content_id: String ID (e.g.'proposed', 'release')
+        :param version: Juju version string for product (i.e. '2.3.3')
+        :param arch: Architecture string of this product (e.g. 's390x','amd64')
+        :param series: Series string that appears in item_name
+          (e.g. 'xenial', 'centos')
+        :param agent_tgz_path: String full path to agent tarball file to use.
+          This file is copied into the JujuStreamData working dir to be served
+          up at a later date.
+        """
+        self.stream_data.add_product(
+            content_id, version, arch, series, agent_tgz_path)
+        # Re-generate when adding a product allows updating the server while
+        # running, can be noisy logs.
+        # self.stream_data.generate_stream_data()
+
+    @contextmanager
+    def server(self):
+        """Serves the products that have been added up until this point.
+
+        :yields: The http address of the server, including the port used.
+        """
+        self.stream_data.generate_stream_data()
+        server = _create_stream_server()
+        ip_address, port = _get_server_address(server)
+        address = 'http://{}:{}'.format(ip_address, port)
+        server_process = multiprocessing.Process(
+            target=_http_worker,
+            args=(server, self.base_dir),
+            name='SimlestreamServer')
+        try:
+            log.info('Starting stream server at: {}'.format(address))
+            multiprocessing.log_to_stderr(logging.DEBUG)
+            server_process.start()
+            yield address
+        finally:
+            log.info('Terminating stream server')
+            server_process.terminate()
+            server_process.join()
+
+
+def agent_tgz_from_juju_binary(juju_bin_path, tmp_dir, series=None):
+    """
+    Create agent tarball with jujud found with provided juju binary.
+
+    Search the location where `juju_bin_path` resides to attempt to find a
+    jujud in the same location.
+
+    :param juju_bin_path: The path to the juju bin in use.
+    :param tmp_dir: Location to store the generated agent file.
+    :param series: String series to use instead of that of the passed binary.
+      Allows one to overwrite the series of the juju client.
+    :returns: String path to generated
+    """
+    _series_lookup = dict(
```

Comment:
Making this more robust to changing ubuntu series would be most ideal. Can we look for win* and set to win2012? Can we assume ubuntu for unknown series?